awx/main/tasks.py

```diff
@@ -1263,7 +1268,12 @@ def build_args(self, job, **kwargs):
         # it doesn't make sense to rely on ansible-playbook's default of using
         # the current user.
         ssh_username = ssh_username or 'root'
-        args = ['ansible-playbook', '-i', self.build_inventory(job, **kwargs)]
+        args = [
+            self.get_path_to_python(job, **kwargs),
+            shutil.which('ansible-playbook'),
```

Comment:
This confuses me, and in my confusion, I wonder if I have done it wrong itself. The resultant abs path will be due to the current environment variables set (specifically PATH). At this point in code, we have build `env` in python working memory, but the actual environment that we're running under is the environment that dispatcher always runs under. Somehow, the result from `instance. ansible_virtualenv_path` needs to affect the output of `shutil.which('ansible-playbook')`, and that seems impossible, because there's no mechanism for data passing between the two.

There was a somewhat arguable case that we could get away with this in inventory imports, because tasks.py lays down the environment, and then calls the management command. So the management command can rely on the OS to return binaries in the virtualenv that tasks.py set up for us. It's still probably wrong in some other way I haven't dived into yet.

Reply:
Well, the assumption here is that in the context of the dispatcher, you're in the *awx* virtualenv, where `ansible` is _not_ installed.  So calling `shutil.which('ansible-playbook')` will yield `/usr/bin/ansible-playbook`.

Reply:
Another alternative is to check to see if `$awx_venv_path/bin/ansible-playbook` exists, and if so, use that:

```diff
+    def get_path_to_ansible(self, instance, executable='ansible-playbook', **kwargs):
+        venv_path = getattr(instance, 'ansible_virtualenv_path', settings.ANSIBLE_VENV>
+        if os.path.exists(venv_path, 'bin', executable):
+            return os.path.join(venv_path, 'bin')
+        return shutil.which(executable)
+
     def build_private_data(self, job, **kwargs):
         '''
         Return SSH private key data (only if stored in DB as ssh_key_data).
@@ -1270,7 +1276,7 @@ class RunJob(BaseTask):
         ssh_username = ssh_username or 'root'
         args = [
             self.get_path_to_python(job, **kwargs),
-            shutil.which('ansible-playbook'),
+            self.get_path_to_ansible(job, 'ansible-playbook', **kwargs),
             '-i',
             self.build_inventory(job, **kwargs)
         ]
```

Reply:
That does seem more correct

> So calling shutil.which('ansible-playbook') will yield /usr/bin/ansible-playbook

exactly. Isn't this a problem? Shouldn't playbooks in the venv Ansible, as your diff here does?